ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java

```diff
@@ -80,30 +102,65 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
     }
   }
 
-  /** Get the next request or block until there is one */
+  /** Get the next request or block until there is one
+   * @return*/
   @Override
-  public NetworkRequest receiveRequest() throws InterruptedException {
-
-    NetworkRequest request = null;
-    while (request == null) {
-      request = requestQueue.take();
-      http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - request.getStartTimeInMs());
-      if (request.equals(EmptyRequest.getInstance())) {
-        logger.debug("Request handler {} received shut down command ", request);
-      } else {
-        DataInputStream stream = new DataInputStream(request.getInputStream());
-        try {
-          // The first 8 bytes is size of the request. TCP implementation uses this size to allocate buffer. See {@link BoundedReceive}
-          // Here we just need to consume it.
-          stream.readLong();
-        } catch (IOException e) {
-          logger.error("Encountered an error while reading length out, close the connection", e);
-          closeConnection(request);
-          request = null;
+  public NetworkRequestBundle receiveRequest() throws InterruptedException {
+
+    NetworkRequestBundle networkRequestBundle;
+    NetworkRequest requestToServe;
+    Collection<NetworkRequest> requestsToDrop;
+    do {
+      networkRequestBundle = networkRequestQueue.take();
+      NetworkRequest unqueuedRequest;
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
+        networkRequestBundle.getRequestsToDrop().add(unqueuedRequest);
```

Comment:
It seems counter-intuitive that you're immediately adding new requests to the queue for requests that may get dropped. Why?